### ISSUE TITLE
[k8s.io] Flake/perma-fail: Federation apiserver [Feature:Federation] Admission control should not be able to create resources if namespace does not exist

### ISSUE THREAD

**Issue by quinton-hoole** [kubernetes/kubernetes#31624]
@nikhiljindal One of these tests introduced in #31139 is ~perma-failing since 8/23.  Could you please take a look?

https://k8s-testgrid.appspot.com/google-gce#gce-federation

**Comment by nikhiljindal** [comment]
Not sure why the UI shows that test as failing, but looking at the logs, that test is passing, its the cluster down process that is failing:

From logs:
```
Ran 15 of 380 Specs in 397.244 seconds
SUCCESS! -- 15 Passed | 0 Failed | 0 Pending | 365 Skipped PASS
```

Error in cluster down:
```
Bringing down cluster
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/zones/us-central1-f/instanceGroupManagers/jenkins-us-central1-f-minion-group].
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/global/instanceTemplates/jenkins-us-central1-f-minion-template].
WARNING: Abbreviated flag [--zone] will be disabled in release 132.0.0, use the full name [--zones].
Updated [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/zones/us-central1-f/instances/jenkins-us-central1-f-master].
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/zones/us-central1-f/instances/jenkins-us-central1-f-master].
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/global/firewalls/jenkins-us-central1-f-master-https].
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/regions/us-central1/addresses/jenkins-us-central1-f-master-ip].
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/global/firewalls/jenkins-us-central1-f-minion-all].
Deleting routes jenkins-us-central1-f-7ee3d5dc-6e06-11e6-bca6-42010af0000a
Deleted [https://www.googleapis.com/compute/v1/projects/k8s-jkns-e2e-gce-federation/global/routes/jenkins-us-central1-f-7ee3d5dc-6e06-11e6-bca6-42010af0000a].
property "clusters.federation-e2e-gce-us-central1-f" unset.
property "users.federation-e2e-gce-us-central1-f" unset.
property "users.federation-e2e-gce-us-central1-f-basic-auth" unset.
property "contexts.federation-e2e-gce-us-central1-f" unset.
property "current-context" unset.
Cleared config for federation-e2e-gce-us-central1-f from /workspace/.kube/config
Done
2016/08/29 09:57:49 e2e.go:451: Step 'teardown' finished in 13m5.463586662s
2016/08/29 09:57:49 e2e.go:185: Sleeping for 30 seconds...
2016/08/29 09:58:19 e2e.go:251: Listing resources...
Listed 0 items.
Listed 0 items.
Listed 0 items.
Listed 0 items.
Listed 0 items.
2016/08/29 09:58:25 e2e.go:206: Created file: /workspace/_artifacts/gcp-resources-before.txt
2016/08/29 09:58:25 e2e.go:206: Created file: /workspace/_artifacts/gcp-resources-cluster-up.txt
2016/08/29 09:58:25 e2e.go:206: Created file: /workspace/_artifacts/gcp-resources-after.txt
2016/08/29 09:58:25 e2e.go:206: Created file: /workspace/_artifacts/gcp-resources-diff.txt
2016/08/29 09:58:25 e2e.go:245: Error: 3 leaked resources
2016/08/29 09:58:25 e2e.go:246: +k8s-fw-acd0a62c46e0711e6b94942010af0000  jenkins-e2e  0.0.0.0/0     tcp:80                                  jenkins-us-central1-a-minion
+acd0a62c46e0711e6b94942010af0000  us-central1  104.154.242.161  TCP          us-central1/targetPools/acd0a62c46e0711e6b94942010af0000
+acd0a62c46e0711e6b94942010af0000  us-central1
exit status 1
+ chmod -R o+r /workspace/_artifacts
+ rc=1
+ [[ 1 -eq 124 ]]
+ [[ 1 -eq 137 ]]
+ [[ 1 -ne 0 ]]
+ echo 'Build failed'
Build failed
+ echo 'Exiting with code: 1'
Exiting with code: 1
+ exit 1
```

**Comment by nikhiljindal** [comment]
cc @ixdy @kubernetes/test-infra-maintainers Are other tests seeing a similar resource leakage error? The script to bring down the cluster is same for all (nothing special for federation).

**Comment by fejta** [comment]
No, other jobs are not leaking.

You can check this by looking at the other tabs in the testgrid link Qutinton provided: https://k8s-testgrid.appspot.com/google-gce#gce

**Comment by madhusudancs** [comment]
@quinton-hoole @nikhiljindal These are the resources leaked by federation. My (narrowed down) suspicion at the moment are either federation-apiserver service or the resources created by federated-services e2e tests which we don't clean up correctly. I will continue to investigate it further.

**Comment by nikhiljindal** [comment]
@madhusudancs pointed me to the diff in artifacts. We are leaking the resources corresponding to the LBs we are creating.

```
@@ -83,0 +84 @@ [ firewall-rules ]
+k8s-fw-aa304627a6e1711e69d8142010af0000  jenkins-e2e  0.0.0.0/0     tcp:80                                  jenkins-us-central1-a-minion
@@ -113,0 +115 @@ [ forwarding-rules ]
+aa304627a6e1711e69d8142010af0000  us-central1  104.197.88.190   TCP          us-central1/targetPools/aa304627a6e1711e69d8142010af0000
@@ -135,0 +138 @@ [ target-pools ]
+aa304627a6e1711e69d8142010af0000  us-central1
```

**Comment by nikhiljindal** [comment]
We need a way to find out if the service controller has deleted LB related cloud resources when we delete our service. Right now, we might kill the controller before it gets a chance to delete those resources, resulting in the leak.
There is no way to find that out using the API right now (https://github.com/kubernetes/kubernetes/issues/32157), but as a temporary measure, we can look for the `DeletedLoadBalancer` event: https://github.com/kubernetes/kubernetes/blob/033e473e12580c73b3bff1319a42b68a61388efd/pkg/controller/service/servicecontroller.go#L788

**Comment by bprashanth** [comment]
You need to poll on the expected resource names in your test, till they disappear or a timeout hits. If a timeout hits, fail the test with "servicecontroller failed to delete cloud resources for %v", but cleanup the resources from the test. See what we do in the ingress e2es: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress.go#L104

**Comment by bprashanth** [comment]
This is a general issue for service e2es, they shutdown too fast and an immediate kube-down will leak. @girishkalele this is what I was describing on your pr, and probably root cause for:  https://github.com/kubernetes/kubernetes/issues/27883

**Comment by nikhiljindal** [comment]
> You need to poll on the expected resource names in your test, till they disappear or a timeout hits. If a timeout hits, fail the test with "servicecontroller failed to delete cloud resources for %v", but cleanup the resources from the test. See what we do in the ingress e2es: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress.go#L104

IIUC, that is not what ingress e2e is doing. ingress e2e always calls cleanupGCE. It is not testing that ingress controller should automatically delete the cloud resources (if given enough time)?

**Comment by goltermann** [comment]
@quinton-hoole @nikhiljindal This hasn't been touched in a while, but is listed in v1.4, can we move to v1.5?

**Comment by quinton-hoole** [comment]
This is actually under active resolution by @madhusudancs, it's just not reflected in this issue.  Yes, it is no longer a v1.4 release blocker.  I will move it to v1.5.

**Comment by quinton-hoole** [comment]
I can take this one @madhusudancs .  Reassigning.

**Comment by quinton-hoole** [comment]
Fixed by #34652
